src/librustc_typeck/collect.rs

```diff
@@ -2603,6 +2603,35 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
             }
         } else if attr.check_name(sym::link_name) {
             codegen_fn_attrs.link_name = attr.value_str();
+        } else if attr.check_name(sym::link_ordinal) {
+            use syntax::ast::{Lit, LitIntType, LitKind};
+            let meta_item_list = attr.meta_item_list();
+            let sole_meta_lit = if let Some(meta_item_list) = &meta_item_list {
+                if meta_item_list.len() == 1 {
+                    meta_item_list.get(0).and_then(|item| item.literal())
+                } else {
+                    None
+                }
+            } else {
+                None
+            };
+            if let Some(Lit { node: LitKind::Int(ordinal, LitIntType::Unsuffixed), .. }) =
+                sole_meta_lit
+            {
+                if *ordinal <= std::usize::MAX as u128 {
+                    codegen_fn_attrs.link_ordinal = Some(*ordinal as usize);
+                } else {
+                    let msg = format!(
+                        "too large ordinal value in link_ordinal \
+                         value: `{}`",
+                        &ordinal
+                    );
+                    tcx.sess.span_err(attr.span, &msg);
+                }
+            } else {
+                let msg = "illegal ordinal format in link_ordinal";
```

Comment:
This can probably be inlined into the method call when introducing `fn check_link_ordinal`.